mlflow/azureml/__init__.py

```diff
@@ -4,139 +4,295 @@
-                        current directory if not specified.
+def _create_execution_script(azure_model):
     """
-    output = os.path.abspath(output)
-    if os.path.exists(output):
-        raise Exception("output folder {} already exists".format(output))
-    os.mkdir(output)
-    if run_id:
-        model_path = _get_model_log_dir(model_path, run_id)
-    model_path = os.path.abspath(model_path)
-    curr_dir = os.path.abspath(os.getcwd())
-    os.chdir(output)
-    try:
-        exec_str = _export("$1", model_path, mlflow_home=mlflow_home)
-        with open("create_service.sh", "w") as f:
-            f.write("\n".join(["#! /bin/sh", "cd {}".format(output), exec_str, ""]))
-    finally:
-        os.chdir(curr_dir)
+    Creates an Azure-compatibele execution script (entry point) for a model server backed by
+    the specified model. This script is created as a temporary file in the current working
+    directory.
 
+    :param registered_model: The Azure Model that the execution script will load for inference.
+    :return: A reference to the temporary file containing the execution script.
+    """
+    # Azure ML requires the container's execution script to be located
+    # in the current working directory during image creation, so we
+    # create the execution script as a temporary file in the current directory
+    execution_script_file = tempfile.NamedTemporaryFile(
+            dir=os.getcwd(), mode="w", prefix="driver", suffix=".py")
+    execution_script_text = SCORE_SRC.format(
+            model_name=azure_model.name, model_version=azure_model.version)
+    execution_script_file.write(execution_script_text)
+    execution_script_file.seek(0)
+    return execution_script_file
+
+
+def _create_dockerfile(output_path, mlflow_path=None):
+    """
+    Creates a Dockerfile containing additional Docker build steps to execute
+    when building the Azure container image. These build steps perform the following tasks:
 
-def _export(app_name, model_path, mlflow_home):
-    conf = _load_conf(model_path)
-    score_py = "score.py"  # NOTE: azure ml requires the main module to be in the current directory
+    - Install MLflow
 
-    with open(score_py, "w") as f:
-        f.write(SCORE_SRC)
+    :param output_path: The path where the Dockerfile will be written.
+    :param mlflow_path: Path to a local copy of the MLflow GitHub repository. If specified, the
+                        Dockerfile command for MLflow installation will install MLflow from this
+                        directory. Otherwise, it will install MLflow from pip.
+    """
+    docker_cmds = ["RUN pip install azureml-sdk"]
 
-    deps = ""
-    mlflow_dep = "mlflow=={}".format(mlflow_version)
+    if mlflow_path is not None:
+        mlflow_install_cmd = "RUN pip install -e {mlflow_path}".format(
+            mlflow_path=_get_container_path(mlflow_path))
+    else:
+        mlflow_install_cmd = "RUN pip install mlflow=={mlflow_version}".format(
```

Comment:
btw, maybe we could somewhere throw an error if we notice mlflow version contains 'dev' and suggest the user use mlflow_home instead.

Reply:
Good idea - Done!